files/zh-cn/web/api/document_object_model/whitespace/index.md

```diff
@@ -1,87 +1,381 @@
+
+{{EmbedLiveSample('HTML 会大幅度忽略空白符吗')}}
+
+这是为了使空白字符不影响你的页面布局。在元素周围和内部创造空间是 CSS 的工作。
+
+### 空白符发生了什么？
+
+然而，它们不会就这样消失。
+
+任何在原始文档中的 HTML 元素之外的空白字符都会在 DOM 中表现出来。这在内部是需要的，以便编辑器可以保留文档的格式。这意味着：
+
+- 会有一些文本节点只包含空白，而且
+- 有些文本节点会在开头或结尾有空白。
+
+以下列文档为例：
+
+```html
+<!DOCTYPE html>
+<html lang="en-US">
+  <head>
+    <meta charset="UTF-8" />
+    <title>My Document</title>
+  </head>
+  <body>
+    <h1>Header</h1>
+    <p>Paragraph</p>
+  </body>
+</html>
+```
+
+其 DOM 树类似于这样：
+
+![代表单个 HTML 文档的 DOM 树](whitespace_tree.png)
+
+在 DOM 中保留空白字符在很多方面都是有用的，但在某些地方，这使得某些布局更难实现，并给那些想在 DOM 中迭代节点的开发者带来了问题。我们将在后面讨论这些问题，以及一些解决方案。
+
+### CSS 如何处理空白符？
+
+大多数空白字符会被忽略，但不是所有的都会被忽略。在前面的示例中，“Hello”和“World!”之间的一个空格在浏览器中渲染时仍然存在。浏览器引擎中有一些规则来决定哪些空白字符是有用的，哪些是没用的——这些规则至少在 [CSS 文本模块第 3 级](https://www.w3.org/TR/css-text-3/)中作了部分说明，特别是关于 [CSS `white-space` 属性](https://www.w3.org/TR/css-text-3/#white-space-property)和[空白符处理细节](https://www.w3.org/TR/css-text-3/#white-space-processing)的部分，但我们也在下面提供一个更简单的解释。
+
+#### 示例
+
+让我们看看其他示例，为了使它看起来更简单，我们添加了一个注释，将所有的空格显示为“◦”，所有的制表符显示为“⇥”，所有的换行符显示为“⏎”：
+
+这个示例
+
+```html-nolint
+<h1>   Hello
+        <span> World!</span>   </h1>
+
+<!--
+<h1>◦◦◦Hello◦⏎
+⇥⇥⇥⇥<span>◦World!</span>⇥◦◦</h1>
+-->
+```
+
+会在浏览器中像这样渲染：
+
+{{EmbedLiveSample('示例')}}
+
+#### 解释
+
+`<h1>` 元素只包含行内元素，实际上包含：
+
+- 一个文本节点（包含一些空格，单词“Hello”和一些制表符）。
+- 一个行内元素（`<span>`，包含一个空格和单词“Hello”）。
+- 另外一个文本节点（只包含制表符和空格）。
+
+正因为如此，它建立了所谓的[行内格式化上下文](/zh-CN/docs/Web/CSS/Inline_formatting_context)。这是浏览器引擎可能使用的布局渲染上下文之一。
+
+在这个上下文中，空白符的处理可以总结如下：
+
+1. 首先，换行前后的所有空格和制表符都会被忽略，所以，如果我们拿之前的示例标记：
+
+   ```html-nolint
+   <h1>◦◦◦Hello◦⏎
+   ⇥⇥⇥⇥<span>◦World!</span>⇥◦◦</h1>
+   ```
+
+   然后应用这条规则，会得到：
+
+   ```html-nolint
+   <h1>◦◦◦Hello⏎
+   <span>◦World!</span>⇥◦◦</h1>
+   ```
+
+2. 然后，所有制表符视作空格字符，所以这个示例会变为：
+
+   ```html-nolint
+   <h1>◦◦◦Hello⏎
+   <span>◦World!</span>◦◦◦</h1>
+   ```
+
+3. 然后，换行符被转化为空格符：
+
+   ```html-nolint
+   <h1>◦◦◦Hello◦<span>◦World!</span>◦◦◦</h1>
+   ```
+
+4. 之后，任何紧跟在另一个空格之后的空格（甚至跨越两个独立的行内元素）都会被忽略，所以我们会得到：
+
+   ```html-nolint
+   <h1>Hello◦<span>World!</span></h1>
+   ```
+
+5. 最后，元素开头和结尾的空格序列被删除，所以我们最终得到了：
+
+   ```html-nolint
+   <h1>Hello◦<span>World!</span></h1>
+   ```
+
+这就是为什么人们在访问网页时，会看到“Hello World!”这句话很好地写在页面的顶部，而不是一个奇怪的缩进的“Hello”，但在下面一行有一个更奇怪的缩进的“World!”。
+
+> **备注：** 从版本 52 开始，[Firefox 开发者工具]支持高亮文本节点，使得观察包含节点中包含的空白字符变得更加容易。纯的空白符节点会有一个“whitespace”标记。
+
+### 在块格式化上下文的空白符
+
+上面我们只是看了包含行内元素的元素，以及行内格式化上下文。如果一个元素至少包含一个块元素，那么它就会建立所谓的[块格式化上下文](zh-CN/docs/Web/Guide/CSS/Block_formatting_context)。
+
+这种情况下，空白符的处理方式非常不同。
+
+#### 示例
+
+让我们看看一个示例来解释发生了什么。我们已经像以前一样标记了空白字符。
+
+我们有三个文本节点，一个在第一个 `<div>` 之前，一个在 2 个 `<div>` 之间，还有一个在第二个 `<div>` 之后。
+
+```html-nolint
 <body>
-  <h1>Header</h1>
-  <p>
-    Paragraph
-  </p>
+  <div>  Hello  </div>
+
+   <div>  World!   </div>
 </body>
-</html>
+
+<!--
+<body>⏎
+⇥<div>◦◦Hello◦◦</div>⏎
+⏎
+◦◦◦<div>◦◦World!◦◦</div>◦◦⏎
+</body>
+-->
 ```
 
-![](whitespace_tree.png)
+它会像这样渲染：
 
-这么一来，要使用 DOM 游走于节点结构间又不想要无用的空白符时，会有点困难。
+{{EmbedLiveSample('示例')}}
```

Comment:
```suggestion
{{EmbedLiveSample('示例_2')}}
```

note that, the duplicate id will be renamed.

![image](https://github.com/mdn/translated-content/assets/15844309/d62f55fe-c98d-4429-855d-c1ff01682e62)